### PR TITLE
Merge of recent updates

### PR DESCRIPTION
**Pull request by jprofijt**
Done:
* Reworked data objects
* Jobs are now based on aggregates from database
* implemented some flex elements
* Added multiple graphs to runtime statistics, new and old
* fixed warning display
* fixed user comments
* added navigation tabs for other graphs
* Styling changes
* Linting errors and warnings fixed

ToDo:
- [x] Write tests for helpers
- [x] Update tests for mutations
- [x] Write and update tests for components

Do in future update:
- write tests for actions
- write more tests for types

### CODE REVIEW

File: src/components/SampleStatisticsComponents/SampleCountsGraph.vue [hunk 1]
```diff
@@ -206,6 +263,44 @@ export default {
           }
         }
       }
+    },
+    /**
+     * Returns date labels for each day of the month. This gives the graph more context into where the data came from
+     * @returns {Record<string<Day>, string<dayOfMonth/Month>} - for examlpe 30th of november = 30/11
```
Comment (jelmerveen) [inline src/components/SampleStatisticsComponents/SampleCountsGraph.vue]: and what happens on 31th of Januari?

File: src/components/TimingStatistics.vue [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+}
+.primary {
+    color: $primary
+}
+.secondary {
+    color: $secondary
+}
+
+.h-10 {
+  height: 5%
```
Comment (jelmerveen) [inline src/components/TimingStatistics.vue]: this makes no sense, how come you don't introduce a new class .h-5?

File: src/components/TimingStatistics.vue [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+}
+
+.h-90 {
+  height: 85%
```
Comment (jelmerveen) [inline src/components/TimingStatistics.vue]: how come you don't make a class .h-85?

File: src/components/Track&Trace-Components/ProjectTimer.vue [hunk 4]
```diff
@@ -1,13 +1,25 @@
...
 import { formatTime, calculateHours, calculateMinutes, calculateSeconds, timeUnit } from '@/helpers/time'
-export default {
+
+declare module 'vue/types/vue' {
+  interface Vue {
+    startTime: number;
+    finishTime: number;
+    started: boolean;
+    time: number;
+    timer: string;
```
Comment (jelmerveen) [inline src/components/Track&Trace-Components/ProjectTimer.vue]: timer suggests that it's a function, but it's actually a string representation. So maybe timeDisplay or timeString would describe it better

File: src/components/TimingStatistics.vue [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+     */
+    types (): string[] {
+      return this.seriesRuntimes.map((type: {name: string; data: {y: number | null; x: string}[]}) => {
+        return type.name
+      })
+    },
+    /**
+     * Builds the series for the median duration times bar chart
+     */
+    axis (): {series: {name: string; data: number[]}[]; xcategories: Array<string | null>} {
```
Comment (jelmerveen) [inline src/components/TimingStatistics.vue]: xcategories? that usggests there are also ycategories?

File: src/components/TimingStatistics.vue (hunk 2 above)
Comment (jelmerveen) [inline src/components/TimingStatistics.vue]: .h-5

File: src/components/Track&Trace-Components/ProgressBar.vue [hunk 6]
```diff
@@ -4,16 +4,29 @@
...
+<script lang="ts">
+import Vue from 'vue'
 
-export default {
+declare module 'vue/types/vue' {
+  interface Vue {
+    step: number;
+    totalSteps: number;
+    noWarning: boolean;
+    varaint: string;
```
Comment (jelmerveen) [inline src/components/Track&Trace-Components/ProgressBar.vue]: varaint?

File: src/components/Track&Trace-Components/RunStatusTable.vue [hunk 7]
```diff
@@ -104,18 +99,45 @@
...
 import helpModalContent from '@/components/Track&Trace-Components/HelpModalContent.vue'
-import { Run, RunStatusData } from '@/types/dataTypes'
+import { RunStatusData } from '@/types/dataTypes'
 import { getFilteredArray } from '@/helpers/utils'
+import { RunData } from '@/types/Run'
+
+declare module 'vue/types/vue' {
+  interface Vue {
+    checkbox: boolean;
+    mouse: string;
```
Comment (jelmerveen) [inline src/components/Track&Trace-Components/RunStatusTable.vue]: mouse is not clear, maybe name it mouseLocation

File: src/components/SampleStatisticsComponents/SampleCountsGraph.vue (hunk 1 above)
Comment (jprofijt) [reply src/components/SampleStatisticsComponents/SampleCountsGraph.vue]: Updated JSdoc to make it clear how and what the function is used for

File: src/components/TimingStatistics.vue (hunk 2 above)
Comment (jprofijt) [reply src/components/TimingStatistics.vue]: removed

File: src/components/TimingStatistics.vue (hunk 3 above)
Comment (jprofijt) [reply src/components/TimingStatistics.vue]: renamed, forgot to rename class after change

File: src/components/Track&Trace-Components/RunTable.vue [hunk 8]
```diff
@@ -162,10 +169,20 @@ export default Vue.extend({
...
-      this.$bvModal.show('comment-modal')
+    openModal (project: string): void {
+      this.selectedProject = ''
+      this.modalLoading = true
+      this.getExtraProjectInfo(project).then(() => {
+          this.modalLoading = false
+          this.selectedProject = project
+          this.$bvModal.show('comment-modal')
+        }).catch(() => {
+          this.$bvToast.toast('Loading of extra project infromation failed', {
```
Comment (jelmerveen) [inline src/components/Track&Trace-Components/RunTable.vue]: typo: infromation > information

File: src/components/Track&Trace-Components/RunTable.vue [hunk 9]
```diff
@@ -188,38 +205,34 @@ export default Vue.extend({
...
-        case pipelineType.exoom:
-          return this.thresholdExoom
-        case pipelineType.pcs:
-          return this.thresholdPcs
-        case pipelineType.svp:
-          return this.thresholdSvp
-        default:
-          return 15
-      }
+      this.loadedProjectInfo[project] ? this.loadedProjectInfo[project].comment = comment : this.loadedProjectInfo[project] = {samples: [], comment: comment}
```
Comment (jelmerveen) [inline src/components/Track&Trace-Components/RunTable.vue]: this line if code is not very readable. Suggestion:

if(!this.loadedProjectInfo[project]) {
    this.loadedProjectInfo[project].samples = []
}

this.loadedProjectInfo[project].comment = comment

File: src/components/Track&Trace-Components/RunTableStepTracker.vue [hunk 10]
```diff
@@ -60,181 +99,301 @@ export default Vue.extend({
...
+     */
+    stepTwoToThreeColor (): string[] {
+      return this.getLineColor(this.pipelinesStatus)
+    },
+    /**
+     * returns the current pipline Running status
+     * 
+     * @returns {stepStatus}
+     */
+    pipelinesStatus (): stepStatus {
```
Comment (jelmerveen) [inline src/components/Track&Trace-Components/RunTableStepTracker.vue]: your  comments suggest that it's a single pipeline, but your code says pipelines

File: src/components/Track&Trace-Components/RunTableStepTracker.vue [hunk 11]
```diff
@@ -60,181 +99,301 @@ export default Vue.extend({
...
+.step-label {
+  z-index: 2;
+  align-self: center;
+  position: absolute;
+  left: -10;
 }
 
-.circle-container li.complete + li:after {
-    background-color: $success;
+.devider {
```
Comment (jelmerveen) [inline src/components/Track&Trace-Components/RunTableStepTracker.vue]: typo: divider

File: src/components/Track&Trace-Components/RunTableStepTracker.vue [hunk 12]
```diff
@@ -60,181 +99,301 @@ export default Vue.extend({
...
+.step-icon {
+  height: 2vw;
+  width: 2vw;
 }
-.progress-text {
-    vertical-align: middle;
+
+.message {
+  font-size: 1vw;
+  border-width: 0.1vw !important;
```
Comment (jelmerveen) [inline src/components/Track&Trace-Components/RunTableStepTracker.vue]: what makes you use !important?

File: src/helpers/time.ts [hunk 13]
```diff
@@ -58,3 +58,5 @@ export function calculateMinutes (milliseconds: number): number {
 export function calculateSeconds (milliseconds: number): number {
   return Math.sqrt(Math.pow(Math.round(((milliseconds / 1000) % 3600) % 60), 2))
```
Comment (jelmerveen) [inline src/helpers/time.ts]: can you explain why you do this?

File: src/store/actions.ts [hunk 14]
```diff
@@ -83,48 +88,42 @@ async function getRunData ({ commit, state: { overviewTable } }: {commit: any, s
...
-        resolve()
-      })
-      .catch(function (error: any) {
-        reject('Could not update jobs')
-      })
+export function findMax (seriesArray: IdentifiedSerie[], length: number): number {
+  if (length === 1) {
+    return seriesArray[0].getLenght()
+  }
+  return max(findMax(seriesArray, length - 1), seriesArray[length - 1].getLenght())
```
Comment (jelmerveen) [inline src/store/actions.ts]: typo: length

you can also do an array function, like sort instead of recursion.

File: src/store/actions.ts [hunk 15]
```diff
@@ -429,19 +408,21 @@ async function handleCommentSubmit ({ dispatch }: { dispatch: any }, { project,
  * @param __namedParameters1.newComment - comment content user wants to change
  * @category TrackAndTrace
  */
-async function checkForCommentUpdates ({ dispatch, state: { projectsTable } }: { dispatch: any, state: State }, { project, oldComment, newComment }: { project: string, oldComment: string, newComment: string }): Promise<string> {
+export async function checkForCommentUpdates ({ commit, dispatch, state: { projectsTable, loadedProjectInfo } }: { commit: (mutation: string, params: object) => void; dispatch: (action: string, params: object) => Promise<object>; state: State }, { project, oldComment, newComment }: { project: string; oldComment: string; newComment: string }): Promise<string> {
   return new Promise((resolve, reject) => {
     api.get(`/api/v1/${projectsTable}/${project}/comment`)
```
Comment (jelmerveen) [inline src/store/actions.ts]: do you use api v1 for a reason?

File: src/store/actions.ts [hunk 16]
```diff
@@ -522,46 +502,227 @@ async function convertProjects ({ commit, state: { projects }, getters: { getJob
...
+ * @returns {Promise<void>}
+ */
+export async function getJobAggregates ({ commit, state: { jobTable } }: {commit: (mutation: string, params: object) => void; state: State}): Promise<void> {
   return new Promise((resolve, reject) => {
-    api.get(`/api/v2/${state.jobTable}?aggs=x==status;y==project;distinct==project_job`)
-      .then((result: {aggs: {matrix: number[][], xLabels: string[], yLabels: string[]}}) => {
-        const projectCounters: Record<string, Record<string, number>> = {}
-        result.aggs.yLabels.forEach((project: string, index:number) => {
-          const statusType: Record<string, number> = {}
+    api.get(`/api/v2/${jobTable}?aggs=x==status;y==project;distinct==project_job`)
```
Comment (jelmerveen) [inline src/store/actions.ts]: and here you use v2

File: src/helpers/time.ts (hunk 13 above)
Comment (jprofijt) [reply src/helpers/time.ts]: short explanation: milliseconds can be negative, but result must be positive.

i could change this to Math.abs(seconds) or force positive input and handle in the components that use still negative numbers.

Longer explanation: ProjectTimer.vue used to be able to act as a countdown timer as well, so negative number inputs had to be possible

Comment (jelmerveen) [reply src/helpers/time.ts]: Math.abs(seconds / 1000) would be in this case better. Nice suggestion!

File: src/types/Run.ts [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+  getStatus(): statusCode;
+  getCurrentStep(): number;
+}
+
+export function constructSteps (demultiplexing: statusCode, raw: statusCode, running: {finished: number; started: number; waiting: number}, copying: {started: number; finished: number; waiting: number}, finished: {total: number; finished: number}): Step[] {
+  const stepOne = new DefaultStep('demultiplexing', demultiplexing)
+  const stepTwo = new DefaultStep('Copying Raw files to PRM', raw)
+  const stepThree = new CounterStep('running pipelines', running.finished + running.started + running.waiting)
+  const stepFour = new CounterStep('copying result files', copying.waiting + copying.started + copying.finished)
+  const stepFive = new DualStateStep('Finished worlflow', finished.total)
```
Comment (jelmerveen) [inline src/types/Run.ts]: typo: workflow

File: src/store/actions.ts (hunk 15 above)
Comment (jprofijt) [reply src/store/actions.ts]: The query only responds with a single value. 
APIv1 response  and query is very straightforward:
`/api/v1/projectsTable/project/comment`
`{ "href": "url", "comment": "Comment text"}`

APIv2:
`/api/v2/projectsTable/project?attrs=comment`
returns metadata, permissions etc + api v1 response

and it was recommended to me for single values 🙃

File: src/store/actions.ts (hunk 16 above)
Comment (jprofijt) [reply src/store/actions.ts]: Here i require api v2 functionality, see above comment

File: src/types/graphTypes.ts [hunk 18]
```diff
@@ -2,127 +2,316 @@
...
  * Stores xAnnotation options
  */
-export interface xAnnotation extends Annotation {
-  x: number
-  x2: number
-  fillColor: string
-  opacity: number
+export interface XaxisAnnotation extends Annotation {
+  x: number;
+  x2: number;
```
Comment (jelmerveen) [inline src/types/graphTypes.ts]: what does x2 do?

File: src/store/actions.ts (hunk 14 above)
Comment (jprofijt) [reply src/store/actions.ts]: I agree.

`seriesArray.reduce((max, number) => { return max < number ? number : max }, 0)`
Is something i could implement.

File: src/types/graphTypes.ts (hunk 18 above)
Comment (jprofijt) [reply src/types/graphTypes.ts]: https://apexcharts.com/docs/options/annotations/

_If you provide this additional x2 value, region will be drawn from x to x2_

draws an annotation region in the graph

File: src/components/Track&Trace-Components/RunTableStepTracker.vue (hunk 10 above)
Comment (jprofijt) [reply src/components/Track&Trace-Components/RunTableStepTracker.vue]: Renamed to projects to be more relevant to the dashboard, which themselves are pipelines.
And updated the comments

File: src/components/Track&Trace-Components/RunTableStepTracker.vue (hunk 12 above)
Comment (jprofijt) [reply src/components/Track&Trace-Components/RunTableStepTracker.vue]: Overwrites bootstrap styles

File: src/store/actions.ts (hunk 15 above)
Comment (jprofijt) [reply src/store/actions.ts]: Api v1 has been removed from actions 😄

File: src/types/graphTypes.ts (hunk 18 above)
Comment (jprofijt) [reply src/types/graphTypes.ts]: Added some links and detail to documentation

**Pull request merged**